### ISSUE TITLE
Server crash for unknown reason

### ISSUE THREAD

**Issue by QiuFeng54321** [itzg/docker-minecraft-server#691]
Using Azure
Used
```bash
docker run -p 25565:25565 -d --name mccontainer -e SERVER_NAME=WilliamCraft -e LEVEL=survival -e MEMORY=2G -e ONLINE_MODE=FALSE -e TYPE=PAPER -e EULA=TRUE -v mccontainer/mcserver:/data itzg/minecraft-server:adopt11
```
to setup
Log: [latest.log](https://github.com/itzg/docker-minecraft-server/files/5701960/latest.log)

**Comment by QiuFeng54321** [comment]
and extra quit message:
```
2020-12-16T10:25:47.773Z	WARN	mc-server-runner	sub-process failed	{"exitCode": -1}
2020-12-16T10:25:47.774Z	INFO	mc-server-runner	Done
```

**Comment by itzg** [comment]
I'm not sure either why it stopped at that point. I wonder if created a crash log that you could provide to the PaperMC team to diagnose?

**Comment by QiuFeng54321** [comment]
I can't find the crash log. There is only a 'logs' folder that exists and I provided the log to you. Is it the crash log you mean?

**Comment by itzg** [comment]
Sometimes it will create one with "crash" in the name, so it sounds like there isn't one.

**Comment by tomtomdurrant** [comment]
I am seeing a similar error message with running FTBA and CURSEFORGE mods, FTBRevealtion3.4 and 3.2 respectively. I'm trying to port my server over to Azure, and am seeing a similar error. 

```
2020-12-19T20:57:30.178Z        WARN    mc-server-runner        sub-process failed      {"exitCode": -1}
```

The mod pack works fine if I run it on my PC locally.

**Comment by al20ov** [comment]
I have a similar issue on multiarch running on a Raspberry Pi 3B (arm64 kernel). Non-multiarch runs fine on my amd64 machine on the same exact version. Here's what `docker logs <my_container>` shows:
```
[15:24:01] [Worker-Main-9/INFO]: Preparing spawn area: 34%
[15:24:02] [Worker-Main-8/INFO]: Preparing spawn area: 34%
[15:24:02] [Worker-Main-9/INFO]: Preparing spawn area: 34%
2020-12-25T15:24:56.993Z	WARN	mc-server-runner	sub-process failed	{"exitCode": -1}
2020-12-25T15:24:58.928Z	INFO	mc-server-runner	Done
```

**Comment by itzg** [comment]
@al20ov in your case I'm guessing the health check is timing out, especially since running the server on RaspberryPi is incredibly slow:

https://github.com/itzg/docker-minecraft-server#healthcheck

@QiuFeng54321 and that makes me realize that maybe the health check is also timing out in your case.

**Comment by aschoettler** [comment]
@QiuFeng54321 Maybe this will help you. [This article](https://www.docker.com/blog/deploying-a-minecraft-docker-server-to-the-cloud/) says to make sure to allocate enough memory in the vm for Minecraft. In the example they have 1g for Minecraft and 1.5G for the azure instance.

**Comment by k-jiang** [comment]
We can confirm that this issue is caused by out of docker's memory. We are running a kubernetes network and we have finally figured it out.

TL;DR: Set your "MEMORY" environment variable 500MB - 1.5GB **less** than your docker memory limit. If you didn't specify your docker's limit while startup, it defaults to 2GB.

Long story:

If you use too much memory than the docker allowed, or your consumption is more than what is available from the system, Docker / Kubernetes will kill your docker.

By default, JVM needs a bit more memory to run besides your "MEMORY" (aka -Xmx) settings. This overhead is for JVM itselfs and it could be between 500MB to 1.5GB depending on your Java version.

Also if you are using JDK 11 with the new G1GC as default, then this number could be higher. In our case we have a bad issue with overhead keep growing uncontrollably, we have to limit it by adding env `JVM_XX_OPTS=-XX:MaxMetaspaceSize=512M`. In JDK 11, the MaxMetaspaceSize is defaults to infinite (yes! [you hear it right](https://stackoverflow.com/questions/62127596/), Java memory management is bull s**t). Even though this extra env alone is good enough, please do note that it only limits **some** of those JVM overheads, and it does not have any effects if you are not using G1GC (e.g. the default Parallel GC from Java 8).

**Comment by itzg** [comment]
In the `MEMORY` section I should really note the discrepancy between JVM heap memory (`MEMORY` variable) and the overall process memory allocation. The latter is indeed what Docker, Kubernetes, etc will enforce.

**Comment by itzg** [comment]
Finally updated wording in MEMORY section:

https://github.com/itzg/docker-minecraft-server#memory-limit
